{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "63d9ccf7",
   "metadata": {},
   "source": [
    "# Score variants of concerns (VOCs) using escape maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04f66bed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "E:\\ESCAPE_MAP_DRAFT\\PGM\\source\\numba_utilities.py:1124: NumbaPerformanceWarning: \u001b[1m\u001b[1mnp.dot() is faster on contiguous arrays, called on (Array(float32, 2, 'F', False, aligned=True), Array(float32, 2, 'A', False, aligned=True))\u001b[0m\u001b[0m\n",
      "  dmean_v_dw = np.dot(s1.T, V)\n",
      "E:\\ESCAPE_MAP_DRAFT\\PGM\\source\\numba_utilities.py:961: NumbaPerformanceWarning: \u001b[1m\u001b[1mnp.dot() is faster on contiguous arrays, called on (Array(float32, 1, 'A', False, aligned=True), Array(float32, 2, 'A', False, aligned=True))\u001b[0m\u001b[0m\n",
      "  mean_V = np.dot(weights, V) / sum_weights\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 29 KD vectors\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\maria\\AppData\\Roaming\\Python\\Python312\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.linear_model import Ridge\n",
    "from scipy.stats import spearmanr\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "sys.path.append(\"../\")\n",
    "from utils import *\n",
    "from global_variables import *\n",
    "from escape_map import *\n",
    "pgm_path = \"PGM/\"\n",
    "sys.path.append(pgm_path + \"source/\")\n",
    "sys.path.append(pgm_path + \"utilities/\")\n",
    "import utilities, Proteins_utils, sequence_logo, plots_utils\n",
    "import rbm, RBM_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "10202ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "VOCS=Proteins_utils.load_FASTA(\"../seq_data/sequences_vocs.fasta\", drop_duplicates=False)\n",
    "VOCS = VOCS[:,BEGIN:-END]\n",
    "def extract_voc_names_from_fasta(fasta_path):\n",
    "    \"\"\"\n",
    "    Extracts VOC names from a FASTA file.\n",
    "    Assumes each entry starts with a '>' followed by the VOC name.\n",
    "    \"\"\"\n",
    "    voc_names = []\n",
    "    with open(fasta_path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                voc_name = line[1:].strip()\n",
    "                voc_names.append(voc_name)\n",
    "    return voc_names\n",
    "\n",
    "VOC_NAMES = extract_voc_names_from_fasta(\"../seq_data/sequences_vocs.fasta\")#viruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "480fce13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         voc       score\n",
      "0      Delta -439.241665\n",
      "1      Alpha -440.173811\n",
      "2         WT -440.639500\n",
      "3    BA.2.75 -447.673339\n",
      "4       BA.2 -454.475632\n",
      "5       BA.1 -464.416719\n",
      "6       BA.5 -465.039454\n",
      "7      Gamma -467.983364\n",
      "8       Beta -473.059786\n",
      "9     BQ.1.1 -473.994005\n",
      "10  NB.1.8.1 -476.527403\n",
      "11      JN.1 -481.806836\n",
      "12       XBB -482.067587\n",
      "13        EG -487.529408\n"
     ]
    }
   ],
   "source": [
    "# --- Build EscapeMap with all concentrations at 10^-12 (ACE2 and all antibodies) ---\n",
    "raw_conc = np.full(len(KD_VECTORS), -5.0, dtype=np.float64)  # antibodies\n",
    "model = EscapeMap(\n",
    "    rbm=RBM,\n",
    "    kd_vectors=KD_VECTORS,\n",
    "    ace2_vector=ACE2_KD_VECTOR,\n",
    "    raw_concentrations=raw_conc,\n",
    "    raw_ace2=-9,      # ACE2 concentration in log10 space\n",
    "    raw_beta=-1,\n",
    "    total_beta=1.0,\n",
    ")\n",
    "\n",
    "# --- Score all VOCs ---\n",
    "n = min(len(VOC_NAMES), VOCS.shape[0])  # guard against mismatch\n",
    "scores = np.array([model.forward(VOCS[i]) for i in range(n)], dtype=float)\n",
    "\n",
    "vocs_df = pd.DataFrame({\n",
    "    \"voc\": VOC_NAMES[:n],\n",
    "    \"score\": scores\n",
    "}).sort_values(\"score\", ascending=False).reset_index(drop=True)\n",
    "\n",
    "print(vocs_df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
